govcd/filter_engine.go

```diff
@@ -0,0 +1,381 @@
+		case FilterParentId:
+			conditions = append(conditions, conditionDef{key, parentIdCondition{value}})
+
+		case FilterLatest:
+			searchLatest = stringToBool(value)
+
+		case FilterEarliest:
+			searchEarliest = stringToBool(value)
+
+		default:
+			return nil, explanation, fmt.Errorf("[SearchByFilter] filter '%s' not supported (only allowed %v), %s)", key, value, supportedFilters)
+		}
+	}
+
+	// We can't allow the search for both the oldest and the newest item
+	if searchEarliest && searchLatest {
+		return nil, explanation, fmt.Errorf("only one of '%s' or '%s' can be used for a set of criteria", FilterEarliest, FilterLatest)
+	}
+
+	var metadataFilter = make(map[string]MetadataFilter)
+	// Fill metadata filters
+	if len(criteria.Metadata) > 0 {
+		for _, cond := range criteria.Metadata {
+			k := cond.Key
+			v := cond.Value
+			isSystem = cond.IsSystem
+			if k == "" {
+				return nil, explanation, fmt.Errorf("metadata condition without key detected")
+			}
+			if v == "" {
+				return nil, explanation, fmt.Errorf("empty value for metadata condition with key '%s'", k)
+			}
+
+			// If we use the metadata search through the API, we must make sure that the type is set
+			if criteria.UseMetadataApiFilter {
+				if cond.Type == "" || strings.EqualFold(cond.Type, "none") {
+					return nil, explanation, fmt.Errorf("requested search by metadata field '%s' must provide a valid type", cond.Key)
+				}
+
+				// The type must be one of the expected values
+				err := validateMetadataType(cond.Type)
+				if err != nil {
+					return nil, explanation, fmt.Errorf("type '%s' for metadata field '%s' is invalid. :%s", cond.Type, cond.Key, err)
+				}
+				metadataFilter[cond.Key] = MetadataFilter{
+					Type:  cond.Type,
+					Value: fmt.Sprintf("%v", cond.Value),
+				}
+			}
+
+			// If we don't use metadata search via the API, we add the field to the list, and
+			// also add a condition, using regular expressions
+			if !criteria.UseMetadataApiFilter {
+				metadataFields = append(metadataFields, k)
+				re, err := regexp.Compile(v.(string))
+				if err != nil {
+					return nil, explanation, fmt.Errorf("error compiling regular expression '%s' : %s ", v, err)
+				}
+				conditions = append(conditions, conditionDef{"metadata", metadataRegexpCondition{k, re}})
+			}
+		}
+	} else {
+		criteria.UseMetadataApiFilter = false
+	}
+
+	var itemResult Results
+	var err error
+
+	if criteria.UseMetadataApiFilter {
+		// This result will not include metadata fields. The query will use metadata parameters to restrict the search
+		itemResult, err = queryByMetadata(queryType, nil, params, metadataFilter, isSystem)
+	} else {
+		// This result includes metadata fields, if they exist.
+		itemResult, err = queryWithMetadataFields(queryType, nil, params, metadataFields, isSystem)
+	}
+
+	if err != nil {
+		return nil, explanation, fmt.Errorf("[SearchByFilter] error retrieving query item list: %s", err)
+	}
+	if os.Getenv("GOVCD_FAIL") == "1" {
```

Comment:
The approach was made this way because the output of these user-defined failures are huge. If we use them always, the log will grow considerably. 
A change that I am considering is to make these logs entries depend on `GOVCD_FAIL=X`.

As I mentioned in my intro presentation, I could completely eliminate this backdoor once we finish the review, or we may decide to convert them into a log-based interface

Reply:
Having them is good. In some situations we need all we can get :)
Another thought about logging. What if we gradually introduce loglevels to `GOVCD_LOG` variable. For example log this output only when `GOVCD_LOG=TRACE`. It wouldn't break anything, but could give flexibility without new variables.

Reply:
Changed to a log-based data inspection. Also explained in `CODE_GUIDELINES`